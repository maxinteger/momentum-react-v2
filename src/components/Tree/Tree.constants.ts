const CLASS_PREFIX = 'md-tree';

const DEFAULTS = {
  IS_RENDERED_FLAT: true,
  EXCLUDE_TREE_ROOT: true,
};

const STYLE = {
  wrapper: `${CLASS_PREFIX}-wrapper`,
  clonedVirtualTreeNode: `${CLASS_PREFIX}-cloned-virtual-tree-node`,
  clonedVirtualTreeNodeATBottom: `${CLASS_PREFIX}-cloned-virtual-tree-node-bottom`,
};

const TREE_NAVIGATION_KEYS = ['ArrowUp', 'ArrowDown', 'ArrowLeft', 'ArrowRight'];

export { CLASS_PREFIX, DEFAULTS, STYLE, TREE_NAVIGATION_KEYS };
