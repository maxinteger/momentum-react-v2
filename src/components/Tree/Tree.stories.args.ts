import { commonStyles } from '../../storybook/helper.stories.argtypes';

export default {
  ...commonStyles,
  shouldNodeFocusBeInset: {
    description:
      'Determines whether the focus around list-items should be inset or outset. This is needed for virtualized lists',
    control: { type: 'boolean' },
    table: {
      type: {
        summary: 'boolean',
      },
      defaultValue: {
        summary: 'false',
      },
    },
  },
  excludeTreeRoot: {
    description: 'Determines if the tree root should be excluded from the tree navigation.',
    control: { type: 'boolean' },
    table: {
      type: {
        summary: 'boolean',
      },
      defaultValue: {
        summary: 'true',
      },
    },
  },
};
