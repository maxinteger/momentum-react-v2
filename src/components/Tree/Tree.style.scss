/**
 * Force overwrite the cloned virtual tree node style
 * To make sure it appear as it should be
 */
.md-tree-cloned-virtual-tree-node {
  // Note: position most be sticky, otherwise VO automatically scroll the node into the view
  // which means the user can not scroll away.
  position: sticky !important;
  top: 0 !important;
  height: 0 !important;
  overflow: hidden !important;
  opacity: 0;
}
