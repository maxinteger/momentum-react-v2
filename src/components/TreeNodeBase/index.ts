import { default as TreeNodeBase } from './TreeNodeBase';
import { Props } from './TreeNodeBase.types';
import * as CONSTANTS from './TreeNodeBase.constants';

export { CONSTANTS as TREE_NODE_BASE_CONSTANTS };

export type TreeNodeBaseProps = Props;

export default TreeNodeBase;
