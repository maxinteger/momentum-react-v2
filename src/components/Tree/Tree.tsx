import React, { FC, useState, useCallback, HTMLAttributes } from 'react';
import classnames from 'classnames';

import { STYLE, DEFAULTS } from './Tree.constants';
import { TreeIdNodeMap, Props, TreeContextValue, TreeNodeId } from './Tree.types';
import {
  convertNestedTree2MappedTree,
  getNextActiveNode,
  toggleTreeNodeRecord,
  TreeContext,
} from './Tree.utils';
import { useKeyboard } from '@react-aria/interactions';
import { isFocusableNode } from '../../utils/navigation';

const Tree: FC<Props> = (props: Props) => {
  const {
    className,
    id,
    style,
    children,
    shouldNodeFocusBeInset,
    treeStructure,
    isRenderedFlat = DEFAULTS.IS_RENDERED_FLAT,
    excludeTreeRoot = DEFAULTS.EXCLUDE_TREE_ROOT,
    setVirtualTreeNodeOpenState,
    ...rest
  } = props;

  if (excludeTreeRoot && !treeStructure.children.length) {
    throw new Error('Tree must have at least one child when excludeTreeRoot is true');
  }

  const [tree, setTree] = useState<TreeIdNodeMap>(convertNestedTree2MappedTree(treeStructure));
  const [activeNodeId, setActiveNodeId] = useState<TreeNodeId>(
    excludeTreeRoot ? treeStructure.children[0].id : treeStructure.id
  );

  const toggleTreeNode = useCallback(
    (id: TreeNodeId, isOpen?: boolean) => {
      const newOpenState = isOpen !== undefined ? isOpen : !tree.get(id).isOpen;
      setVirtualTreeNodeOpenState?.(id, newOpenState);
      setTree((prevTree) => toggleTreeNodeRecord(id, prevTree, newOpenState));
    },
    [tree]
  );

  const getNodeDetails = useCallback((id: TreeNodeId) => tree.get(id), [tree]);

  const getNodeProps = useCallback(
    (id: TreeNodeId): Partial<HTMLAttributes<HTMLElement>> => {
      const node = tree.get(id);
      const parent = node.parent && tree.get(node.parent);
      const isRoot = parent === undefined;

      // tabindex depends on the treeNodeBase params as well
      const props = {
        id: node.id.toString(),
        'aria-setsize': isRoot ? 1 : parent.children.length,
        'aria-level': node.level + 1,
        'aria-posinset': node.index + 1,
        role: 'treeitem',
      };
      if (node.isOpen !== undefined) {
        props['aria-expanded'] = node.isOpen.toString();
      }
      return props;
    },
    [tree]
  );

  const getNodeGroupProps = useCallback(
    (id: TreeNodeId): Partial<HTMLAttributes<HTMLElement>> => {
      const node = tree.get(id);

      return {
        role: 'group',
        'aria-owns': node.children.join(' '),
      };
    },
    [tree]
  );

  const getContext = useCallback(
    (): TreeContextValue => ({
      getNodeProps,
      getNodeGroupProps,
      getNodeDetails,
      isRenderedFlat,
      shouldNodeFocusBeInset,
      activeNodeId,
      setActiveNodeId,
      toggleTreeNode,
    }),
    [
      activeNodeId,
      getNodeGroupProps,
      getNodeProps,
      getNodeDetails,
      isRenderedFlat,
      shouldNodeFocusBeInset,
      toggleTreeNode,
    ]
  );

  const { keyboardProps } = useKeyboard({
    onKeyDown: (evt) => {
      const target = evt.target as HTMLElement;
      switch (evt.key) {
        case 'Escape':
          evt.continuePropagation();
          break;
        case 'ArrowUp':
        case 'ArrowDown':
        case 'ArrowRight':
        case 'ArrowLeft':
        case 'Enter': {
          // Ignore event when the target is an interactable element inside the tree node
          if (target.getAttribute('role') === 'treeitem' || !isFocusableNode(target)) {
            evt.preventDefault();
            setActiveNodeId(
              getNextActiveNode(tree, activeNodeId, evt.key, excludeTreeRoot, toggleTreeNode)
            );
          }
          break;
        }
      }
    },
  });

  return (
    <TreeContext.Provider value={getContext()}>
      <div
        className={classnames(className, STYLE.wrapper)}
        style={style}
        id={id}
        role="tree"
        {...keyboardProps}
        {...rest}
      >
        {children}
      </div>
    </TreeContext.Provider>
  );
};

export default Tree;
