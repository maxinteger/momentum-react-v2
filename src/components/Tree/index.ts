import { default as Tree } from './Tree';
import * as CONSTANTS from './Tree.constants';
import { Props } from './Tree.types';

export { CONSTANTS as TREE_CONSTANTS };

export { useTreeContext } from './Tree.utils';

export type TreeProps = Props;
export type { TreeNode, TreeNodeId, TreeRoot, TreeContextValue } from './Tree.types';

export default Tree;
