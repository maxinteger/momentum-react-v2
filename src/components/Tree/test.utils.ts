import { TreeNode, TreeNodeId } from './Tree.types';

/**
 * Create a tree node with the given id, isOpen and children.
 *
 * @param id The id of the node
 * @param isOpenByDefault
 * @param children List of child nodes
 */
export const createTreeNode = (
  id: TreeNodeId,
  isOpenByDefault: boolean = undefined,
  children: Array<TreeNode> = []
): TreeNode => ({
  children,
  isOpenByDefault,
  id,
});
