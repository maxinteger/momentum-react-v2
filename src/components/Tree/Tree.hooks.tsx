import { useEffect } from 'react';

import { STYLE, TREE_NAVIGATION_KEYS } from './Tree.constants';
import { UseVirtualTreeNavigationProps } from './Tree.types';
import './Tree.style.scss';
import { getKeyboardFocusableElements } from '../../utils/navigation';
import { NODE_ID_ATTRIBUTE_NAME, NODE_ID_DATA_NAME } from '../TreeNodeBase/TreeNodeBase.constants';

/**
 * Handle DOM changes for virtual tree
 *
 * @param props
 * @internal
 */
export const useVirtualTreeNavigation = ({
  virtualTreeConnector,
  treeRef,
  activeNodeId,
}: UseVirtualTreeNavigationProps): void => {
  // Handle DOM changes for virtual tree
  useEffect(() => {
    if (!virtualTreeConnector) return;

    // target = parent of the first tree node to avoid mutation change detection in the subtree
    const targetNode = treeRef.current.querySelector(`[${NODE_ID_ATTRIBUTE_NAME}]`)?.parentElement;

    if (!targetNode) return;

    // To re-dispatch arrow keydown events
    let keyDownEvent: KeyboardEvent;

    // Remove all cloned nodes
    const cleanUp = () =>
      Array.from(targetNode.querySelectorAll(`.${STYLE.clonedVirtualTreeNode}`)).forEach((e) =>
        e.remove()
      );

    // Filter element by active node id
    const getByNodeId = (node: HTMLElement) => node.dataset[NODE_ID_DATA_NAME] === activeNodeId;

    // Mutation observer to handle the focus change
    const mutationHandler: MutationCallback = (mutationList) => {
      for (const mutation of mutationList) {
        // Handle removed active node
        const removed = Array.from(mutation.removedNodes).find(getByNodeId) as HTMLElement;
        if (removed) {
          // Clone the active node and make some modifications
          const clonedNode = removed.cloneNode(true) as HTMLElement;
          // remove nodeid attribute
          clonedNode.removeAttribute(NODE_ID_ATTRIBUTE_NAME);
          clonedNode.classList.add(STYLE.clonedVirtualTreeNode);
          // make all focusable elements un-focusable
          getKeyboardFocusableElements(clonedNode).forEach((el) =>
            el.setAttribute('tabindex', '-1')
          );
          targetNode.appendChild(clonedNode);
          clonedNode.focus({ preventScroll: true });

          // add focus and keydown event listeners
          clonedNode.addEventListener('focus', () => {
            virtualTreeConnector.scrollToNode(activeNodeId);
          });
          clonedNode.addEventListener('keydown', (evt) => {
            if (TREE_NAVIGATION_KEYS.includes(evt.key) || (evt.key === 'Tab' && !evt.shiftKey)) {
              virtualTreeConnector.scrollToNode(activeNodeId);
              keyDownEvent = new KeyboardEvent('keydown', evt);
              evt.stopPropagation();
              evt.preventDefault();
            }
          });
          return;
        }

        // Handle adding back the focused node
        const added = Array.from(mutation.addedNodes).find(getByNodeId) as HTMLElement;
        if (added) {
          cleanUp();
          added.focus({ preventScroll: true });

          // Handle keydown event originated from the cloned node
          if (keyDownEvent) {
            // re-dispatch Tab key event does not change focus in every browser, so
            // manually focus on the first element
            if (keyDownEvent.key === 'Tab') {
              getKeyboardFocusableElements(added)[0]?.focus?.();
            } else {
              // re-dispatch the arrow key events
              added.dispatchEvent(keyDownEvent);
            }
            keyDownEvent = null;
          }
          return;
        }
      }
    };

    // Start observing the target node
    const observer = new MutationObserver(mutationHandler);
    observer.observe(targetNode, { childList: true });

    // Clean up
    return () => {
      observer.disconnect();
      cleanUp();
    };
  }, [virtualTreeConnector, activeNodeId, treeRef.current]);
};
