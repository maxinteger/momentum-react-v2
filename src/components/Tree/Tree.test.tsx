import React from 'react';
import { mount } from 'enzyme';
import '@testing-library/jest-dom';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import Tree, { TREE_CONSTANTS } from './index';
import { createTreeNode as tNode } from './test.utils';
import TreeNodeBase from '../TreeNodeBase';
import { convertNestedTree2MappedTree, mapTree } from './Tree.utils';

const getSampleTree = () => {
  // prettier-ignore
  return tNode('root', true, [
    tNode('1', false, [
      tNode('1.1',true, [
        tNode('1.1.1'),
        tNode('1.1.2')]),
      tNode('1.2')]),
    tNode('2', true, [
      tNode('2.1'),
      tNode('2.2', true, [
        tNode('2.2.1'),
        tNode('2.2.2'),
        tNode('2.2.3')]),
    ]),
    tNode('3'),
    tNode('4', true, [
      tNode('4.1', false, [
        tNode('4.1.1'),
        tNode('4.1.2')]),
      tNode('4.2')]),
  ]);
};

describe('<Tree />', () => {
  const commonProps = {
    treeStructure: tNode('root', true, [tNode('1'), tNode('2')]),
  };

  describe('snapshot', () => {
    it('should match snapshot', () => {
      expect.assertions(1);

      const container = mount(<Tree {...commonProps} />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with className', () => {
      expect.assertions(1);

      const className = 'example-class';

      const container = mount(<Tree className={className} {...commonProps} />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with id', () => {
      expect.assertions(1);

      const id = 'example-id';

      const container = mount(<Tree id={id} {...commonProps} />);
      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with style', () => {
      expect.assertions(1);

      const style = { color: 'pink' };

      const container = mount(<Tree style={style} {...commonProps} />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with style', () => {
      expect.assertions(1);

      const style = { color: 'pink' };

      const container = mount(<Tree style={style} {...commonProps} />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with style', () => {
      expect.assertions(1);

      const style = { color: 'pink' };

      const container = mount(<Tree style={style} {...commonProps} />);

      expect(container).toMatchSnapshot();
    });
  });

  describe('attributes', () => {
    it('should have its wrapper class', () => {
      expect.assertions(1);

      const container = mount(<Tree {...commonProps} />);
      const element = container.find(Tree).getDOMNode();

      expect(element.classList.contains(TREE_CONSTANTS.STYLE.wrapper)).toBe(true);
    });

    it('should have provided class when className is provided', () => {
      expect.assertions(1);

      const className = 'example-class';
      const container = mount(<Tree className={className} {...commonProps} />);
      const element = container.find(Tree).getDOMNode();

      expect(element.classList.contains(className)).toBe(true);
    });

    it('should have provided id when id is provided', () => {
      expect.assertions(1);

      const id = 'example-id';
      const container = mount(<Tree id={id} {...commonProps} />);
      const element = container.find(Tree).getDOMNode();

      expect(element.id).toBe(id);
    });

    it('should have provided style when style is provided', () => {
      expect.assertions(1);

      const style = { color: 'pink' };
      const styleString = 'color: pink;';
      const container = mount(<Tree style={style} {...commonProps} />);
      const element = container.find(Tree).getDOMNode();

      expect(element.getAttribute('style')).toBe(styleString);
    });

    it('should have provided aria-label when aria-label is provided', () => {
      expect.assertions(1);

      const label = 'test';
      const container = mount(<Tree aria-label={label} {...commonProps} />);
      const element = container.find(Tree).getDOMNode();

      expect(element.getAttribute('aria-label')).toBe('test');
    });

    it('should have provided aria-labelledby when aria-labelledby is provided', () => {
      expect.assertions(1);

      const labelBy = 'label-id';
      const container = mount(<Tree aria-labelledby={labelBy} {...commonProps} />);
      const element = container.find(Tree).getDOMNode();

      expect(element.getAttribute('aria-labelledby')).toBe(labelBy);
    });
  });

  describe('tree focus', () => {
    it('should handle escape being pressed', async () => {
      const keyDownHandler = jest.fn();

      const { getByRole } = render(
        // eslint-disable-next-line jsx-a11y/no-static-element-interactions
        <div onKeyDown={keyDownHandler}>
          <Tree treeStructure={getSampleTree()} excludeTreeRoot={false}>
            <TreeNodeBase key="0" nodeId="root">
              TreeNodeBase 1
            </TreeNodeBase>
          </Tree>
        </div>
      );

      const treeNode = getByRole('treeitem');

      await userEvent.tab();
      expect(treeNode).toHaveFocus();

      await userEvent.keyboard('{Escape}');
      expect(keyDownHandler).toHaveBeenCalled();
    });

    it('should handle up/down arrow keys correctly', async () => {
      expect.assertions(25);
      const user = userEvent.setup();
      const tree = getSampleTree();

      const { getByTestId } = render(
        <Tree treeStructure={tree} excludeTreeRoot={false}>
          {mapTree(
            convertNestedTree2MappedTree(tree),
            (node) => (
              <TreeNodeBase
                key={node.id.toString()}
                nodeId={node.id}
                data-testid={`node-${node.id}`}
              >
                {node.id}
              </TreeNodeBase>
            ),
            { excludeRootNode: false }
          )}
        </Tree>
      );

      await user.tab();

      expect(getByTestId('node-root')).toHaveFocus();

      const nodeOrder = [
        'node-1',
        'node-2',
        'node-2.1',
        'node-2.2',
        'node-2.2.1',
        'node-2.2.2',
        'node-2.2.3',
        'node-3',
        'node-4',
        'node-4.1',
        'node-4.2',
        'node-4.2',
      ];
      for (const testid of nodeOrder) {
        await user.keyboard('{ArrowDown}');
        expect(getByTestId(testid)).toHaveFocus();
      }

      const reversNodeOrder = [
        'node-4.1',
        'node-4',
        'node-3',
        'node-2.2.3',
        'node-2.2.2',
        'node-2.2.1',
        'node-2.2',
        'node-2.1',
        'node-2',
        'node-1',
        'node-root',
        'node-root',
      ];

      for (const testid of reversNodeOrder) {
        await user.keyboard('{ArrowUp}');
        expect(getByTestId(testid)).toHaveFocus();
      }
    });

    it('should handle left/right arrow keys correctly', async () => {
      expect.assertions(13);
      const user = userEvent.setup();
      const tree = getSampleTree();
      const setVirtualTreeNodeOpenState = jest.fn();

      const { getByTestId } = render(
        <Tree
          treeStructure={tree}
          excludeTreeRoot={false}
          setVirtualTreeNodeOpenState={setVirtualTreeNodeOpenState}
        >
          {mapTree(
            convertNestedTree2MappedTree(tree),
            (node) => (
              <TreeNodeBase key={node.id.toString()} nodeId={node.id} data-testid={node.id}>
                {node.id}
              </TreeNodeBase>
            ),
            { excludeRootNode: false }
          )}
        </Tree>
      );

      await user.tab();

      expect(getByTestId('root')).toHaveFocus();

      const nodeOrder = [
        { nextFocusedNode: '1', toggleNode: false },
        { nextFocusedNode: '1', toggleNode: true }, // opens the node
        { nextFocusedNode: '1.1', toggleNode: false },
        { nextFocusedNode: '1.1.1', toggleNode: false },
      ];
      for (const { nextFocusedNode, toggleNode } of nodeOrder) {
        await user.keyboard('{ArrowRight}');
        expect(getByTestId(nextFocusedNode)).toHaveFocus();

        if (toggleNode) {
          expect(setVirtualTreeNodeOpenState).toHaveBeenCalledWith(nextFocusedNode, true);
        }
      }

      const reversNodeOrder = [
        { nextFocusedNode: '1.1', toggleNode: false }, // moved up
        { nextFocusedNode: '1.1', toggleNode: true }, // close
        { nextFocusedNode: '1', toggleNode: false }, // moved up
        { nextFocusedNode: '1', toggleNode: true }, // close
        { nextFocusedNode: 'root', toggleNode: false }, // moved up
      ];

      for (const { nextFocusedNode, toggleNode } of reversNodeOrder) {
        await user.keyboard('{ArrowLeft}');
        expect(getByTestId(nextFocusedNode)).toHaveFocus();

        if (toggleNode) {
          expect(setVirtualTreeNodeOpenState).toHaveBeenCalledWith(nextFocusedNode, false);
        }
      }
    });
  });
});
