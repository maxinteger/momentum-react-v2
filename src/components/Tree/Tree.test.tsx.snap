// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tree /> snapshot should match snapshot 1`] = `
<ForwardRef
  treeStructure={
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "id": "1",
          "isOpenByDefault": undefined,
        },
        Object {
          "children": Array [],
          "id": "2",
          "isOpenByDefault": undefined,
        },
      ],
      "id": "root",
      "isOpenByDefault": true,
    }
  }
>
  <div
    className="md-tree-wrapper"
    onKeyDown={[Function]}
    role="tree"
  />
</ForwardRef>
`;

exports[`<Tree /> snapshot should match snapshot with className 1`] = `
<ForwardRef
  className="example-class"
  treeStructure={
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "id": "1",
          "isOpenByDefault": undefined,
        },
        Object {
          "children": Array [],
          "id": "2",
          "isOpenByDefault": undefined,
        },
      ],
      "id": "root",
      "isOpenByDefault": true,
    }
  }
>
  <div
    className="example-class md-tree-wrapper"
    onKeyDown={[Function]}
    role="tree"
  />
</ForwardRef>
`;

exports[`<Tree /> snapshot should match snapshot with id 1`] = `
<ForwardRef
  id="example-id"
  treeStructure={
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "id": "1",
          "isOpenByDefault": undefined,
        },
        Object {
          "children": Array [],
          "id": "2",
          "isOpenByDefault": undefined,
        },
      ],
      "id": "root",
      "isOpenByDefault": true,
    }
  }
>
  <div
    className="md-tree-wrapper"
    id="example-id"
    onKeyDown={[Function]}
    role="tree"
  />
</ForwardRef>
`;

exports[`<Tree /> snapshot should match snapshot with style 1`] = `
<ForwardRef
  style={
    Object {
      "color": "pink",
    }
  }
  treeStructure={
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "id": "1",
          "isOpenByDefault": undefined,
        },
        Object {
          "children": Array [],
          "id": "2",
          "isOpenByDefault": undefined,
        },
      ],
      "id": "root",
      "isOpenByDefault": true,
    }
  }
>
  <div
    className="md-tree-wrapper"
    onKeyDown={[Function]}
    role="tree"
    style={
      Object {
        "color": "pink",
      }
    }
  />
</ForwardRef>
`;

exports[`<Tree /> snapshot should match snapshot with style 2`] = `
<ForwardRef
  style={
    Object {
      "color": "pink",
    }
  }
  treeStructure={
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "id": "1",
          "isOpenByDefault": undefined,
        },
        Object {
          "children": Array [],
          "id": "2",
          "isOpenByDefault": undefined,
        },
      ],
      "id": "root",
      "isOpenByDefault": true,
    }
  }
>
  <div
    className="md-tree-wrapper"
    onKeyDown={[Function]}
    role="tree"
    style={
      Object {
        "color": "pink",
      }
    }
  />
</ForwardRef>
`;

exports[`<Tree /> snapshot should match snapshot with style 3`] = `
<ForwardRef
  style={
    Object {
      "color": "pink",
    }
  }
  treeStructure={
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "id": "1",
          "isOpenByDefault": undefined,
        },
        Object {
          "children": Array [],
          "id": "2",
          "isOpenByDefault": undefined,
        },
      ],
      "id": "root",
      "isOpenByDefault": true,
    }
  }
>
  <div
    className="md-tree-wrapper"
    onKeyDown={[Function]}
    role="tree"
    style={
      Object {
        "color": "pink",
      }
    }
  />
</ForwardRef>
`;
